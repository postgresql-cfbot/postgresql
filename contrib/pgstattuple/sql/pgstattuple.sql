CREATE EXTENSION pgstattuple;
--
-- It's difficult to come up with platform-independent test cases for
-- the pgstattuple functions, but the results for empty tables and
-- indexes should be that.
--

create table test (a int primary key, b int[], p point) with (autovacuum_enabled = off);

insert into test(a, b, p)
select a, array(select generate_series as num from generate_series(1, 10)), point(a*10, a*10) from generate_series(1, 10000) a;
-- make dead tuples
delete from test where a between 1 and 5000;

select * from pgstattuple('test');
select * from pgstattuple('test'::text);
select * from pgstattuple('test'::name);
select * from pgstattuple('test'::regclass);
select pgstattuple(oid) from pg_class where relname = 'test';
select pgstattuple(relname) from pg_class where relname = 'test';
select * from pgstattuple_approx('test');
select * from pgstattuple_approx('test'::text);
select * from pgstattuple_approx('test'::regclass);

select version, tree_level,
    index_size / current_setting('block_size')::int as index_size,
    root_block_no, internal_pages, leaf_pages, empty_pages, deleted_pages,
    avg_leaf_density, leaf_fragmentation
    from pgstatindex('test_pkey');
select version, tree_level,
    index_size / current_setting('block_size')::int as index_size,
    root_block_no, internal_pages, leaf_pages, empty_pages, deleted_pages,
    avg_leaf_density, leaf_fragmentation
    from pgstatindex('test_pkey'::text);
select version, tree_level,
    index_size / current_setting('block_size')::int as index_size,
    root_block_no, internal_pages, leaf_pages, empty_pages, deleted_pages,
    avg_leaf_density, leaf_fragmentation
    from pgstatindex('test_pkey'::name);
select version, tree_level,
    index_size / current_setting('block_size')::int as index_size,
    root_block_no, internal_pages, leaf_pages, empty_pages, deleted_pages,
    avg_leaf_density, leaf_fragmentation
    from pgstatindex('test_pkey'::regclass);

select pg_relpages('test');
select pg_relpages('test_pkey');
select pg_relpages('test_pkey'::text);
select pg_relpages('test_pkey'::name);
select pg_relpages('test_pkey'::regclass);
select pg_relpages(oid) from pg_class where relname = 'test_pkey';
select pg_relpages(relname) from pg_class where relname = 'test_pkey';

create index test_ginidx on test using gin (b);

select * from pgstatginindex('test_ginidx');

create index test_hashidx on test using hash (b);

select * from pgstathashindex('test_hashidx');

-- check that (btree, hash, gist) index with pgstattuple should work
create index test_btreeidx on test using btree (a);
create index test_gistidx on test using gist (p);

select * from pgstattuple('test_btreeidx');
select * from pgstattuple('test_hashidx');
select * from pgstattuple('test_gistidx');

-- check that these index type error with (gin, spgist, brin, etc)
create index test_spgistidx on test using spgist (p);
create index test_brinidx on test using brin (a);

select * from pgstattuple('test_ginidx');
select * from pgstattuple('test_spgistidx');
select * from pgstattuple('test_brinidx');

-- these should error with the wrong type
select pgstatginindex('test_pkey');
select pgstathashindex('test_pkey');

select pgstatindex('test_ginidx');
select pgstathashindex('test_ginidx');

select pgstatindex('test_hashidx');
select pgstatginindex('test_hashidx');

-- check that using any of these functions with unsupported relations will fail
create table test_partitioned (a int) partition by range (a);
create index test_partitioned_index on test_partitioned(a);
-- these should all fail
select pgstattuple('test_partitioned');
select pgstattuple('test_partitioned_index');
select pgstattuple_approx('test_partitioned');
select pg_relpages('test_partitioned');
select pgstatindex('test_partitioned');
select pgstatginindex('test_partitioned');
select pgstathashindex('test_partitioned');

create view test_view as select 1;
-- these should all fail
select pgstattuple('test_view');
select pgstattuple_approx('test_view');
select pg_relpages('test_view');
select pgstatindex('test_view');
select pgstatginindex('test_view');
select pgstathashindex('test_view');

create foreign data wrapper dummy;
create server dummy_server foreign data wrapper dummy;
create foreign table test_foreign_table () server dummy_server;
-- these should all fail
select pgstattuple('test_foreign_table');
select pgstattuple_approx('test_foreign_table');
select pg_relpages('test_foreign_table');
select pgstatindex('test_foreign_table');
select pgstatginindex('test_foreign_table');
select pgstathashindex('test_foreign_table');

-- a partition of a partitioned table should work though
create table test_partition partition of test_partitioned for values from (1) to (100);
select pgstattuple('test_partition');
select pgstattuple_approx('test_partition');
select pg_relpages('test_partition');

-- toast tables should work
select pgstattuple((select reltoastrelid from pg_class where relname = 'test'));
select pgstattuple_approx((select reltoastrelid from pg_class where relname = 'test'));
select pg_relpages((select reltoastrelid from pg_class where relname = 'test'));

-- not for the index calls though, of course
select pgstatindex('test_partition');
select pgstatginindex('test_partition');
select pgstathashindex('test_partition');

-- an actual index of a partitioned table should work though
create index test_partition_idx on test_partition(a);
create index test_partition_hash_idx on test_partition using hash (a);
-- these should work
select pgstatindex('test_partition_idx');
select pgstathashindex('test_partition_hash_idx');

drop table test_partitioned;
drop view test_view;
drop foreign table test_foreign_table;
drop server dummy_server;
drop foreign data wrapper dummy;
