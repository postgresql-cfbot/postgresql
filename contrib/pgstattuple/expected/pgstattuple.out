CREATE EXTENSION pgstattuple;
--
-- It's difficult to come up with platform-independent test cases for
-- the pgstattuple functions, but the results for empty tables and
-- indexes should be that.
--
create table test (a int primary key, b int[], p point) with (autovacuum_enabled = off);
insert into test(a, b, p)
select a, array(select generate_series as num from generate_series(1, 10)), point(a*10, a*10) from generate_series(1, 10000) a;
-- make dead tuples
delete from test where a between 1 and 5000;
select * from pgstattuple('test');
 table_len | tuple_count | tuple_len | tuple_percent | dead_tuple_count | dead_tuple_len | dead_tuple_percent | free_space | free_percent 
-----------+-------------+-----------+---------------+------------------+----------------+--------------------+------------+--------------
   1171456 |        5000 |    560000 |          47.8 |             5000 |         560000 |               47.8 |       7452 |         0.64
(1 row)

select * from pgstattuple('test'::text);
 table_len | tuple_count | tuple_len | tuple_percent | dead_tuple_count | dead_tuple_len | dead_tuple_percent | free_space | free_percent 
-----------+-------------+-----------+---------------+------------------+----------------+--------------------+------------+--------------
   1171456 |        5000 |    560000 |          47.8 |             5000 |         560000 |               47.8 |       7452 |         0.64
(1 row)

select * from pgstattuple('test'::name);
 table_len | tuple_count | tuple_len | tuple_percent | dead_tuple_count | dead_tuple_len | dead_tuple_percent | free_space | free_percent 
-----------+-------------+-----------+---------------+------------------+----------------+--------------------+------------+--------------
   1171456 |        5000 |    560000 |          47.8 |             5000 |         560000 |               47.8 |       7452 |         0.64
(1 row)

select * from pgstattuple('test'::regclass);
 table_len | tuple_count | tuple_len | tuple_percent | dead_tuple_count | dead_tuple_len | dead_tuple_percent | free_space | free_percent 
-----------+-------------+-----------+---------------+------------------+----------------+--------------------+------------+--------------
   1171456 |        5000 |    560000 |          47.8 |             5000 |         560000 |               47.8 |       7452 |         0.64
(1 row)

select pgstattuple(oid) from pg_class where relname = 'test';
                      pgstattuple                      
-------------------------------------------------------
 (1171456,5000,560000,47.8,5000,560000,47.8,7452,0.64)
(1 row)

select pgstattuple(relname) from pg_class where relname = 'test';
                      pgstattuple                      
-------------------------------------------------------
 (1171456,5000,560000,47.8,5000,560000,47.8,7452,0.64)
(1 row)

select * from pgstattuple_approx('test');
 table_len | scanned_percent | approx_tuple_count | approx_tuple_len | approx_tuple_percent | dead_tuple_count | dead_tuple_len | dead_tuple_percent | approx_free_space | approx_free_percent 
-----------+-----------------+--------------------+------------------+----------------------+------------------+----------------+--------------------+-------------------+---------------------
   1171456 |             100 |               5000 |           560000 |    47.80375874125874 |             5000 |         560000 |  47.80375874125874 |              7452 |  0.6361314466783217
(1 row)

select * from pgstattuple_approx('test'::text);
 table_len | scanned_percent | approx_tuple_count | approx_tuple_len | approx_tuple_percent | dead_tuple_count | dead_tuple_len | dead_tuple_percent | approx_free_space | approx_free_percent 
-----------+-----------------+--------------------+------------------+----------------------+------------------+----------------+--------------------+-------------------+---------------------
   1171456 |             100 |               5000 |           560000 |    47.80375874125874 |             5000 |         560000 |  47.80375874125874 |              7452 |  0.6361314466783217
(1 row)

select * from pgstattuple_approx('test'::regclass);
 table_len | scanned_percent | approx_tuple_count | approx_tuple_len | approx_tuple_percent | dead_tuple_count | dead_tuple_len | dead_tuple_percent | approx_free_space | approx_free_percent 
-----------+-----------------+--------------------+------------------+----------------------+------------------+----------------+--------------------+-------------------+---------------------
   1171456 |             100 |               5000 |           560000 |    47.80375874125874 |             5000 |         560000 |  47.80375874125874 |              7452 |  0.6361314466783217
(1 row)

select version, tree_level,
    index_size / current_setting('block_size')::int as index_size,
    root_block_no, internal_pages, leaf_pages, empty_pages, deleted_pages,
    avg_leaf_density, leaf_fragmentation
    from pgstatindex('test_pkey');
 version | tree_level | index_size | root_block_no | internal_pages | leaf_pages | empty_pages | deleted_pages | avg_leaf_density | leaf_fragmentation 
---------+------------+------------+---------------+----------------+------------+-------------+---------------+------------------+--------------------
       4 |          1 |         30 |             3 |              1 |         28 |           0 |             0 |            87.91 |                  0
(1 row)

select version, tree_level,
    index_size / current_setting('block_size')::int as index_size,
    root_block_no, internal_pages, leaf_pages, empty_pages, deleted_pages,
    avg_leaf_density, leaf_fragmentation
    from pgstatindex('test_pkey'::text);
 version | tree_level | index_size | root_block_no | internal_pages | leaf_pages | empty_pages | deleted_pages | avg_leaf_density | leaf_fragmentation 
---------+------------+------------+---------------+----------------+------------+-------------+---------------+------------------+--------------------
       4 |          1 |         30 |             3 |              1 |         28 |           0 |             0 |            87.91 |                  0
(1 row)

select version, tree_level,
    index_size / current_setting('block_size')::int as index_size,
    root_block_no, internal_pages, leaf_pages, empty_pages, deleted_pages,
    avg_leaf_density, leaf_fragmentation
    from pgstatindex('test_pkey'::name);
 version | tree_level | index_size | root_block_no | internal_pages | leaf_pages | empty_pages | deleted_pages | avg_leaf_density | leaf_fragmentation 
---------+------------+------------+---------------+----------------+------------+-------------+---------------+------------------+--------------------
       4 |          1 |         30 |             3 |              1 |         28 |           0 |             0 |            87.91 |                  0
(1 row)

select version, tree_level,
    index_size / current_setting('block_size')::int as index_size,
    root_block_no, internal_pages, leaf_pages, empty_pages, deleted_pages,
    avg_leaf_density, leaf_fragmentation
    from pgstatindex('test_pkey'::regclass);
 version | tree_level | index_size | root_block_no | internal_pages | leaf_pages | empty_pages | deleted_pages | avg_leaf_density | leaf_fragmentation 
---------+------------+------------+---------------+----------------+------------+-------------+---------------+------------------+--------------------
       4 |          1 |         30 |             3 |              1 |         28 |           0 |             0 |            87.91 |                  0
(1 row)

select pg_relpages('test');
 pg_relpages 
-------------
         143
(1 row)

select pg_relpages('test_pkey');
 pg_relpages 
-------------
          30
(1 row)

select pg_relpages('test_pkey'::text);
 pg_relpages 
-------------
          30
(1 row)

select pg_relpages('test_pkey'::name);
 pg_relpages 
-------------
          30
(1 row)

select pg_relpages('test_pkey'::regclass);
 pg_relpages 
-------------
          30
(1 row)

select pg_relpages(oid) from pg_class where relname = 'test_pkey';
 pg_relpages 
-------------
          30
(1 row)

select pg_relpages(relname) from pg_class where relname = 'test_pkey';
 pg_relpages 
-------------
          30
(1 row)

create index test_ginidx on test using gin (b);
select * from pgstatginindex('test_ginidx');
 version | pending_pages | pending_tuples 
---------+---------------+----------------
       2 |             0 |              0
(1 row)

create index test_hashidx on test using hash (b);
select * from pgstathashindex('test_hashidx');
 version | bucket_pages | overflow_pages | bitmap_pages | unused_pages | live_items | dead_items |   free_percent    
---------+--------------+----------------+--------------+--------------+------------+------------+-------------------
       4 |           17 |             12 |            1 |           15 |       5000 |          0 | 72.12061736104916
(1 row)

-- check that (btree, hash, gist) index with pgstattuple should work
create index test_btreeidx on test using btree (a);
create index test_gistidx on test using gist (p);
select * from pgstattuple('test_btreeidx');
 table_len | tuple_count | tuple_len | tuple_percent | dead_tuple_count | dead_tuple_len | dead_tuple_percent | free_space | free_percent 
-----------+-------------+-----------+---------------+------------------+----------------+--------------------+------------+--------------
    131072 |        5000 |     80000 |         61.04 |                0 |              0 |                  0 |      13812 |        10.54
(1 row)

select * from pgstattuple('test_hashidx');
ERROR:  index "test_hashidx" contains unexpected zero page at block 31
HINT:  Please REINDEX it.
select * from pgstattuple('test_gistidx');
 table_len | tuple_count | tuple_len | tuple_percent | dead_tuple_count | dead_tuple_len | dead_tuple_percent | free_space | free_percent 
-----------+-------------+-----------+---------------+------------------+----------------+--------------------+------------+--------------
    237568 |        5000 |    200000 |         84.19 |                0 |              0 |                  0 |       8144 |         3.43
(1 row)

-- check that these index type error with (gin, spgist, brin, etc)
create index test_spgistidx on test using spgist (p);
create index test_brinidx on test using brin (a);
select * from pgstattuple('test_ginidx');
ERROR:  index "test_ginidx" (gin index) is not supported
select * from pgstattuple('test_spgistidx');
ERROR:  index "test_spgistidx" (spgist index) is not supported
select * from pgstattuple('test_brinidx');
ERROR:  index "test_brinidx" (brin index) is not supported
-- these should error with the wrong type
select pgstatginindex('test_pkey');
ERROR:  relation "test_pkey" is not a GIN index
select pgstathashindex('test_pkey');
ERROR:  relation "test_pkey" is not a hash index
select pgstatindex('test_ginidx');
ERROR:  relation "test_ginidx" is not a btree index
select pgstathashindex('test_ginidx');
ERROR:  relation "test_ginidx" is not a hash index
select pgstatindex('test_hashidx');
ERROR:  relation "test_hashidx" is not a btree index
select pgstatginindex('test_hashidx');
ERROR:  relation "test_hashidx" is not a GIN index
-- check that using any of these functions with unsupported relations will fail
create table test_partitioned (a int) partition by range (a);
create index test_partitioned_index on test_partitioned(a);
-- these should all fail
select pgstattuple('test_partitioned');
ERROR:  cannot get tuple-level statistics for relation "test_partitioned"
DETAIL:  This operation is not supported for partitioned tables.
select pgstattuple('test_partitioned_index');
ERROR:  cannot get tuple-level statistics for relation "test_partitioned_index"
DETAIL:  This operation is not supported for partitioned indexes.
select pgstattuple_approx('test_partitioned');
ERROR:  relation "test_partitioned" is of wrong relation kind
DETAIL:  This operation is not supported for partitioned tables.
select pg_relpages('test_partitioned');
ERROR:  cannot get page count of relation "test_partitioned"
DETAIL:  This operation is not supported for partitioned tables.
select pgstatindex('test_partitioned');
ERROR:  relation "test_partitioned" is not a btree index
select pgstatginindex('test_partitioned');
ERROR:  relation "test_partitioned" is not a GIN index
select pgstathashindex('test_partitioned');
ERROR:  "test_partitioned" is not an index
create view test_view as select 1;
-- these should all fail
select pgstattuple('test_view');
ERROR:  cannot get tuple-level statistics for relation "test_view"
DETAIL:  This operation is not supported for views.
select pgstattuple_approx('test_view');
ERROR:  relation "test_view" is of wrong relation kind
DETAIL:  This operation is not supported for views.
select pg_relpages('test_view');
ERROR:  cannot get page count of relation "test_view"
DETAIL:  This operation is not supported for views.
select pgstatindex('test_view');
ERROR:  relation "test_view" is not a btree index
select pgstatginindex('test_view');
ERROR:  relation "test_view" is not a GIN index
select pgstathashindex('test_view');
ERROR:  "test_view" is not an index
create foreign data wrapper dummy;
create server dummy_server foreign data wrapper dummy;
create foreign table test_foreign_table () server dummy_server;
-- these should all fail
select pgstattuple('test_foreign_table');
ERROR:  cannot get tuple-level statistics for relation "test_foreign_table"
DETAIL:  This operation is not supported for foreign tables.
select pgstattuple_approx('test_foreign_table');
ERROR:  relation "test_foreign_table" is of wrong relation kind
DETAIL:  This operation is not supported for foreign tables.
select pg_relpages('test_foreign_table');
ERROR:  cannot get page count of relation "test_foreign_table"
DETAIL:  This operation is not supported for foreign tables.
select pgstatindex('test_foreign_table');
ERROR:  relation "test_foreign_table" is not a btree index
select pgstatginindex('test_foreign_table');
ERROR:  relation "test_foreign_table" is not a GIN index
select pgstathashindex('test_foreign_table');
ERROR:  "test_foreign_table" is not an index
-- a partition of a partitioned table should work though
create table test_partition partition of test_partitioned for values from (1) to (100);
select pgstattuple('test_partition');
     pgstattuple     
---------------------
 (0,0,0,0,0,0,0,0,0)
(1 row)

select pgstattuple_approx('test_partition');
  pgstattuple_approx   
-----------------------
 (0,0,0,0,0,0,0,0,0,0)
(1 row)

select pg_relpages('test_partition');
 pg_relpages 
-------------
           0
(1 row)

-- toast tables should work
select pgstattuple((select reltoastrelid from pg_class where relname = 'test'));
     pgstattuple     
---------------------
 (0,0,0,0,0,0,0,0,0)
(1 row)

select pgstattuple_approx((select reltoastrelid from pg_class where relname = 'test'));
  pgstattuple_approx   
-----------------------
 (0,0,0,0,0,0,0,0,0,0)
(1 row)

select pg_relpages((select reltoastrelid from pg_class where relname = 'test'));
 pg_relpages 
-------------
           0
(1 row)

-- not for the index calls though, of course
select pgstatindex('test_partition');
ERROR:  relation "test_partition" is not a btree index
select pgstatginindex('test_partition');
ERROR:  relation "test_partition" is not a GIN index
select pgstathashindex('test_partition');
ERROR:  "test_partition" is not an index
-- an actual index of a partitioned table should work though
create index test_partition_idx on test_partition(a);
create index test_partition_hash_idx on test_partition using hash (a);
-- these should work
select pgstatindex('test_partition_idx');
         pgstatindex          
------------------------------
 (4,0,8192,0,0,0,0,0,NaN,NaN)
(1 row)

select pgstathashindex('test_partition_hash_idx');
   pgstathashindex   
---------------------
 (4,8,0,1,0,0,0,100)
(1 row)

drop table test_partitioned;
drop view test_view;
drop foreign table test_foreign_table;
drop server dummy_server;
drop foreign data wrapper dummy;
